File: src/Api/Producers/ProducerPrograms.cs
using System.Text.Json;
using Api.Decision;

namespace Api.Producers;

public static class ProducerPrograms
{
    static readonly IReadOnlyList<ProducerProgram> Programs = Load();

    static IReadOnlyList<ProducerProgram> Load()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "data", "producer-programs.json");
        return JsonSerializer.Deserialize<List<ProducerProgram>>(File.ReadAllText(path))!;
    }

    public static IReadOnlyList<string> FindCandidates(Assessment assessment)
    {
        var producer = TextHelpers.Normalize(assessment.Brand ?? "");
        var text = TextHelpers.Normalize(string.Join(" ", new[]
        {
            assessment.ObjectName ?? "",
            assessment.Category ?? "",
            assessment.Subcategory ?? "",
        }));

        return Programs
            .Where(program => producer == program.Producer)
            .Where(program => program.EligibleHints.Any(text.Contains)
                           && !program.ExcludedHints.Any(text.Contains))
            .Select(program => program.Id)
            .ToList();
    }

    static IReadOnlyList<ProducerProgram> FindForContext(DecisionContext context)
    {
        var producer = context.Producer;
        if (string.IsNullOrEmpty(producer) || producer is "unknown" or "ved ikke" or "no" or "anden")
            return [];

        return Programs.Where(program => program.Producer == producer).ToList();
    }

    public static ProducerProgramResult Evaluate(DecisionContext context)
    {
        var programs = FindForContext(context);
        if (programs.Count == 0)
        {
            return new ProducerProgramResult
            {
                Status = "none",
                Title = "Producentordninger",
                Message = "Der er ikke fundet en konkret producentordning endnu. "
                        + "I næste fase kan modulet slå op i en database med reparation, "
                        + "reservedele, buy-back, trade-in, take-back og refurbishment.",
                Programs = [],
            };
        }

        var evaluated = programs.Select(program => EvaluateSingle(program, context)).ToList();
        var best = evaluated.OrderByDescending(item => item.Rank).First();

        return new ProducerProgramResult
        {
            Status = best.Status,
            Title = "Producentordning fundet",
            Message = best.Message,
            Programs = evaluated,
        };
    }

    static ProgramEvaluation EvaluateSingle(ProducerProgram program, DecisionContext context)
    {
        var text = string.Join(" ", new[] { context.ObjectName, context.Category, context.Subcategory });
        var hasEligibleHint = program.EligibleHints.Any(text.Contains);
        var hasExcludedHint = program.ExcludedHints.Any(text.Contains);
        var goodCondition = context.Works == "yes" && context.Damage is "no" or "minor";
        var complete = context.Accessories is "complete" or "irrelevant" or null;

        var original = TriState(context.OriginalProduct);
        var clean = TriState(context.CleanState);
        var unmodified = TriState(context.Unmodified);
        var assembled = TriState(context.Assembled);
        bool?[] eligibilityChecks = [original, clean, unmodified, assembled];

        var likely = hasEligibleHint
            && !hasExcludedHint
            && goodCondition
            && complete
            && eligibilityChecks.All(value => value is true);

        var possible = !hasExcludedHint
            && context.Works is "yes" or "unknown"
            && !eligibilityChecks.Any(value => value is false);

        var (status, rank, message) = likely
            ? ("likely", 3, $"{program.Title} ser ud til potentielt at være relevant. "
                + "Hent en officiel vurdering og sammenlign med almindeligt privat salg.")
            : possible
                ? ("possible", 2, $"{program.Title} kan muligvis være relevant, men stand, kategori, "
                    + "original mærkning og komplethed skal bekræftes hos producenten.")
                : ("unlikely", 1, $"{program.Title} ligner ikke en oplagt vej ud fra svarene. "
                    + "Fortsæt med almindelig salgs- eller bortgivelsesvurdering.");

        return new ProgramEvaluation
        {
            Id = program.Id,
            Title = program.Title,
            Producer = program.Producer.ToUpperInvariant(),
            SchemeTypes = program.SchemeTypes,
            Status = status,
            Rank = rank,
            Message = message,
            Url = program.Url,
            Reward = program.Reward,
            SourceLabel = program.SourceLabel,
            Checks =
            [
                new() { Label = "Originalt producentprodukt", Ok = original },
                new() { Label = "God/salgbar stand", Ok = goodCondition },
                new() { Label = "Rent", Ok = clean },
                new() { Label = "Komplet", Ok = complete },
                new() { Label = "Uændret", Ok = unmodified },
                new() { Label = "Korrekt samlet", Ok = assembled },
                new() { Label = "Mulig omfattet kategori", Ok = hasEligibleHint && !hasExcludedHint },
            ],
            Comparison =
            [
                new() { Label = program.Title, Value = program.Reward },
                new() { Label = "Privat salg", Value = "ofte højere pris, mere arbejde" },
                new() { Label = "Hurtigt salg", Value = "lavere pris, hurtigere afhændelse" },
            ],
        };
    }

    static bool? TriState(string? value) => value switch
    {
        "yes" => true,
        "no" => false,
        _ => null,
    };
}

File: legacy/producer_programs.py
PRODUCER_PROGRAMS = [
    {
        "id": "ikea_gensalg",
        "producer": "ikea",
        "title": "IKEA Gensalg",
        "scheme_types": ["buy_back", "resale"],
        "url": "https://www.ikea.com/dk/da/second-hand/sell-to-ikea/quote/",
        "reward": "IKEA tilgodebevis",
        "source_label": "IKEA Gensalg vurderingsværktøj",
        "eligible_hints": [
            "kommode",
            "skaenk",
            "vitrineskab",
            "bogreol",
            "reol",
            "garderobeskab",
            "multimediemoebel",
            "spisebord",
            "skrivebord",
            "sofabord",
            "sengebord",
            "stol",
            "kontorstol",
            "taburet",
            "sengestel",
            "underseng",
            "lamelbund",
            "entre",
            "hylde",
            "opbevaring",
            "bordlampe",
            "gulvlampe",
            "belysning",
            "ovnfast",
            "glas",
            "porcelaen",
            "boernemoebel",
            "baby",
        ],
        "excluded_hints": [
            "hvidevare",
            "elektrisk",
            "elektronik",
            "madras",
            "sengetekstil",
            "havemoebel",
            "udendoers",
            "koekken",
            "bordplade",
            "front",
            "loes del",
        ],
        "checks": [
            "Originalt producentprodukt",
            "God/salgbar stand",
            "Rent og uændret",
            "Komplet og fuldt funktionelt",
            "Korrekt samlet",
            "Omfattet produktkategori",
        ],
    }
]


def find_program_candidates(assessment):
    producer = normalize(str(assessment.get("brand") or ""))
    text = normalize(
        " ".join(
            str(assessment.get(key) or "")
            for key in ("object_name", "category", "subcategory")
        )
    )
    candidates = []
    for program in PRODUCER_PROGRAMS:
        if producer != program["producer"]:
            continue
        eligible = any(hint in text for hint in program["eligible_hints"])
        excluded = any(hint in text for hint in program["excluded_hints"])
        if eligible and not excluded:
            candidates.append(program["id"])
    return candidates


def find_producer_programs(context):
    producer = context.get("producer") or ""
    if not producer or producer in ("unknown", "ved ikke", "no", "anden"):
        return []

    return [program for program in PRODUCER_PROGRAMS if program["producer"] == producer]


def evaluate_producer_program(context):
    programs = find_producer_programs(context)
    if not programs:
        return {
            "status": "none",
            "title": "Producentordninger",
            "message": (
                "Der er ikke fundet en konkret producentordning endnu. "
                "I næste fase kan modulet slå op i en database med reparation, "
                "reservedele, buy-back, trade-in, take-back og refurbishment."
            ),
            "programs": [],
        }

    evaluated = [evaluate_single_program(program, context) for program in programs]
    best = sorted(evaluated, key=lambda item: item["rank"], reverse=True)[0]

    return {
        "status": best["status"],
        "title": "Producentordning fundet",
        "message": best["message"],
        "programs": evaluated,
    }


def evaluate_single_program(program, context):
    text = " ".join(
        [context.get("object_name", ""), context.get("category", ""), context.get("subcategory", "")]
    )
    has_eligible_hint = any(hint in text for hint in program["eligible_hints"])
    has_excluded_hint = any(hint in text for hint in program["excluded_hints"])
    good_condition = context.get("works") == "yes" and context.get("damage") in ("no", "minor", None)
    complete = context.get("accessories") in ("complete", "irrelevant", None)
    original = tri_state(context.get("original_product"))
    clean = tri_state(context.get("clean_state"))
    unmodified = tri_state(context.get("unmodified"))
    assembled = tri_state(context.get("assembled"))

    eligibility_checks = [original, clean, unmodified, assembled]
    likely = (
        has_eligible_hint
        and not has_excluded_hint
        and good_condition
        and complete
        and all(value is True for value in eligibility_checks)
    )
    possible = (
        not has_excluded_hint
        and context.get("works") in ("yes", "unknown")
        and not any(value is False for value in eligibility_checks)
    )

    if likely:
        status = "likely"
        rank = 3
        message = (
            f"{program['title']} ser ud til potentielt at være relevant. "
            "Hent en officiel vurdering og sammenlign med almindeligt privat salg."
        )
    elif possible:
        status = "possible"
        rank = 2
        message = (
            f"{program['title']} kan muligvis være relevant, men stand, kategori, "
            "original mærkning og komplethed skal bekræftes hos producenten."
        )
    else:
        status = "unlikely"
        rank = 1
        message = (
            f"{program['title']} ligner ikke en oplagt vej ud fra svarene. "
            "Fortsæt med almindelig salgs- eller bortgivelsesvurdering."
        )

    return {
        "id": program["id"],
        "title": program["title"],
        "producer": program["producer"].upper(),
        "scheme_types": program["scheme_types"],
        "status": status,
        "rank": rank,
        "message": message,
        "url": program["url"],
        "reward": program["reward"],
        "source_label": program["source_label"],
        "checks": [
            {"label": "Originalt producentprodukt", "ok": original},
            {"label": "God/salgbar stand", "ok": good_condition},
            {"label": "Rent", "ok": clean},
            {"label": "Komplet", "ok": complete},
            {"label": "Uændret", "ok": unmodified},
            {"label": "Korrekt samlet", "ok": assembled},
            {"label": "Mulig omfattet kategori", "ok": has_eligible_hint and not has_excluded_hint},
        ],
        "comparison": [
            {"label": program["title"], "value": program["reward"]},
            {"label": "Privat salg", "value": "ofte højere pris, mere arbejde"},
            {"label": "Hurtigt salg", "value": "lavere pris, hurtigere afhændelse"},
        ],
    }


def tri_state(value):
    if value == "yes":
        return True
    if value == "no":
        return False
    return None


def normalize(value):
    return value.lower().replace("æ", "ae").replace("ø", "oe").replace("å", "aa")
